android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/ComponentWorkerThread.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.android.component.core;
+
+import android.util.Log;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.aws.greengrass.android.component.utils.Constants.EXIT_CODE_FAILED;
+import static com.aws.greengrass.android.component.utils.Constants.EXIT_CODE_TERMINATED;
+
+public abstract class ComponentWorkerThread extends Thread {
+
+    private static final String LOG_TAG = ComponentWorkerThread.class.getSimpleName();
+    private final Semaphore criticalSectionLock = new Semaphore(1, true);
+
+    private AtomicBoolean enableThreadInterrupt = new AtomicBoolean(false);
+
+    /**
+     * Worker thread exit code.
+     * Don't confuse it with service exit code. This code indicates the result of the worker thread.
+     * Upper level Service object may use it if the component is operating normally, but the Service
+     * may also disregard this exit code if the Service itself caught an exception or failed in
+     * other way.
+     */
+    protected int workerExitCode = EXIT_CODE_FAILED; // assume error by default
+
+    /**
+     * Gets the exit code of the worker thread.
+     * @return null if the thread is still running, integer value of the exit code otherwise
+     */
+    public Integer getExitCode() {
+        if (isAlive()) {
+            return null;
+        } else {
+            return workerExitCode;
+        }
+    }
+
+    /**
+     * Don't use this method to stop the thread, use stopWorker() instead.
+     */
+    @Override
+    public final void interrupt() {
+        if (enableThreadInterrupt.getAndSet(false)) {
+            super.interrupt();
+        } else {
+            Log.w(LOG_TAG, ComponentWorkerThread.class.getSimpleName()
+                    + " cannot be interrupted directly. Use stopWorker() method instead");
+        }
+    }
+
+    /**
+     * Stops the worker thread.
+     * Use it instead of interrupt() to gracefully terminate the worker thread
+     */
+    public final void stopWorker() {
+        if (isAlive()) {
+            enterCriticalSection();
+            onStopWorker();
+            enableThreadInterrupt.set(true);
+            interrupt();
+            try {
+                join();
+            } catch (InterruptedException ignore) {

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)